bbq/compiler/compiler.go

```diff
@@ -1472,10 +1458,33 @@ func (c *Compiler[_, _]) loadTypeArguments(expression *ast.InvocationExpression)
 
 func (c *Compiler[_, _]) VisitMemberExpression(expression *ast.MemberExpression) (_ struct{}) {
 	c.compileExpression(expression.Expression)
+
+	memberAccessInfo, ok := c.ExtendedElaboration.MemberExpressionMemberAccessInfo(expression)
+	if !ok {
+		panic(errors.NewUnreachableError())
+	}
+
+	if memberAccessInfo.IsOptional {
+		c.codeGen.Emit(opcode.InstructionUnwrap{})
+	}
+
 	constant := c.addStringConst(expression.Identifier.Identifier)
+
+	// TODO: remove member if `isNestedResourceMove`
+	//  See `Interpreter.memberExpressionGetterSetter` for the reference implementation.
 	c.codeGen.Emit(opcode.InstructionGetField{
 		FieldNameIndex: constant.index,
 	})
+
+	// Return a reference, if the member is accessed via a reference.
+	// This is pre-computed at the checker.
+	if memberAccessInfo.ReturnReference {
+		index := c.getOrAddType(memberAccessInfo.ResultingType)
+		c.codeGen.Emit(opcode.InstructionNewRef{TypeIndex: index})
+	}
```

Comment:
This is the only thing needed to support the  behaviour for entitlement-mappings on references. Its nice that the type-checker does most of the work on computing the result types.

This also works in the same way in the interpreter. i.e: This now matches the `memberExpressionGetterSetter ` of the interpreter: https://github.com/onflow/cadence/blob/285de336ba10bcd929c83307e620a9c525db188a/interpreter/interpreter_expression.go#L245
Left the TODOs for the unimplemented functionalities. We would get to them once we start enabling more tests.